### PR TITLE
Cache cloud provider node instances in cluster state.

### PR DESCRIPTION
**Pull request by krzysztof-jastrzebski**
Cache cloud provider node instances in cluster state.

### CODE REVIEW

File: cluster-autoscaler/clusterstate/clusterstate.go [hunk 1]
```diff
@@ -55,6 +56,11 @@ const (
 
 	// NodeGroupBackoffResetTimeout is the time after last failed scale-up when the backoff duration is reset.
 	NodeGroupBackoffResetTimeout = 3 * time.Hour
+
+	// CloudProviderNodeInstancesCacheRefreshInterval is the interval between cache refreshes.
```
Comment (losipiuk) [inline cluster-autoscaler/clusterstate/clusterstate.go]: between nodegroup instances cache refreshes.

File: cluster-autoscaler/clusterstate/clusterstate.go [hunk 2]
```diff
@@ -55,6 +56,11 @@ const (
...
+	CloudProviderNodeInstancesCacheRefreshInterval = 2 * time.Minute
```
Comment (losipiuk) [inline cluster-autoscaler/clusterstate/clusterstate.go]: CloudProviderNodeInstancesCacheRefreshInterval

File: cluster-autoscaler/clusterstate/clusterstate.go [hunk 3]
```diff
@@ -55,6 +56,11 @@ const (
...
+	// CloudProviderNodeInstancesCacheEntryFreshnessInterval is the duration when cache entry is fresh.
+	CloudProviderNodeInstancesCacheEntryFreshnessInterval = 3 * time.Minute
```
Comment (losipiuk) [inline cluster-autoscaler/clusterstate/clusterstate.go]: CloudProviderNodeInstancesCacheEntryFreshnessThreshold?

File: cluster-autoscaler/clusterstate/clusterstate.go [hunk 4]
```diff
@@ -1090,3 +1115,114 @@ func fakeNode(instance cloudprovider.Instance) *apiv1.Node {
...
+	}
+	return results, nil
+}
+
+// GetCloudProviderNodeInstancesForNodeGroup returns cloud provider node instances for the given node group.
+func (cache *CloudProviderNodeInstancesCache) GetCloudProviderNodeInstancesForNodeGroup(nodeGroup cloudprovider.NodeGroup) ([]cloudprovider.Instance, error) {
+	cacheEntry, found := cache.getCacheEntryLocked(nodeGroup)
+	if found {
+		if cacheEntry.isStale() {
+			klog.Warningf("Entry cloudProviderNodeInstances is stale, refresh time is %v", cacheEntry.refreshTime)
```
Comment (losipiuk) [inline cluster-autoscaler/clusterstate/clusterstate.go]: If we are doing refresh every two minutes and freshness interval is 3 minutes we will be seeing a lot of those.  If we dozens/hundreds of node groups we will not manage to refresh all within 1 minut (3-1)

File: cluster-autoscaler/clusterstate/clusterstate.go [hunk 5]
```diff
@@ -1090,3 +1115,114 @@ func fakeNode(instance cloudprovider.Instance) *apiv1.Node {
...
+	}
+	cache.updateCacheEntryLocked(nodeGroup, &cloudProviderNodeInstancesCacheEntry{nodeGroupInstances, time.Now()})
+	return nodeGroupInstances, nil
+
+}
+
+// InvalidateCacheEntry removes entry for the given node group from cache.
+func (cache *CloudProviderNodeInstancesCache) InvalidateCacheEntry(nodeGroup cloudprovider.NodeGroup) {
+	klog.V(5).Infof("Invalidate entry in cloud provider node instances cache %v", nodeGroup.Id())
+	cache.removeCacheEntryLocked(nodeGroup)
```
Comment (losipiuk) [inline cluster-autoscaler/clusterstate/clusterstate.go]: I am thinking that maybe instead removing entry here we can just make it stale.
Then we can change the refresh algorithm in such way that we are running it more often (say every 10 seconds) and it refreshes only stale entries?

Then if we have many scale-ups in progress we still will not wait for all to be refreshed every loop. But we will prioritize that those are refreshed more often.

I am thinking about the case when we have multiple GPU-enabled nodegroups for which scale-ups take a lot of time (as GPU nodes are slow to come up). Then we can end up with each loop still taking lots of time. 

But maybe this is premature optimization. WDYT?

File: cluster-autoscaler/clusterstate/clusterstate.go [hunk 6]
```diff
@@ -1090,3 +1115,114 @@ func fakeNode(instance cloudprovider.Instance) *apiv1.Node {
...
+}
+
+// Refresh refreshes cache.
+func (cache *CloudProviderNodeInstancesCache) Refresh() {
+	klog.Infof("Start refreshing cloud provider node instances cache")
+	defer klog.Infof("Refresh cloud provider node instances cache finished")
```
Comment (losipiuk) [inline cluster-autoscaler/clusterstate/clusterstate.go]: nit: You can log how much it took

File: cluster-autoscaler/clusterstate/clusterstate_test.go [hunk 7]
```diff
@@ -781,3 +782,68 @@ func TestIsNodeStillStarting(t *testing.T) {
 func newBackoff() backoff.Backoff {
 	return backoff.NewIdBasedExponentialBackoff(InitialNodeGroupBackoffDuration, MaxNodeGroupBackoffDuration, NodeGroupBackoffResetTimeout)
 }
+
+func Test(t *testing.T) {
```
Comment (losipiuk) [inline cluster-autoscaler/clusterstate/clusterstate_test.go]: Please name it somewhat more specific ;)

File: cluster-autoscaler/main.go [hunk 8]
```diff
@@ -314,6 +314,9 @@ func run(healthCheck *metrics.HealthCheck) {
 	// Start updating health check endpoint.
 	healthCheck.StartMonitoring()
 
+	// Start components running in background.
+	autoscaler.Start()
```
Comment (losipiuk) [inline cluster-autoscaler/main.go]: Either handle error here or drop it from signature

File: cluster-autoscaler/clusterstate/clusterstate.go [hunk 9]
```diff
@@ -1090,3 +1115,114 @@ func fakeNode(instance cloudprovider.Instance) *apiv1.Node {
 		},
 	}
 }
+
```
Comment (losipiuk) [inline cluster-autoscaler/clusterstate/clusterstate.go]: please move cache to separate file.

File: cluster-autoscaler/clusterstate/clusterstate.go (hunk 5 above)
Comment (krzysztof-jastrzebski) [reply cluster-autoscaler/clusterstate/clusterstate.go]: I see following problem with such solution: We need to refresh all instances every x minutes. If there is a lot of instance to be refreshed during one refresh then we might end up waiting longer than 10 seconds for refreshing scaled up node groups. Other node groups can starve scaled up node groups for some time.

File: cluster-autoscaler/clusterstate/clusterstate.go (hunk 1 above)
Comment (krzysztof-jastrzebski) [reply cluster-autoscaler/clusterstate/clusterstate.go]: done

File: cluster-autoscaler/clusterstate/clusterstate.go (hunk 2 above)
Comment (krzysztof-jastrzebski) [reply cluster-autoscaler/clusterstate/clusterstate.go]: What needs to be changed

File: cluster-autoscaler/clusterstate/clusterstate.go (hunk 3 above)
Comment (krzysztof-jastrzebski) [reply cluster-autoscaler/clusterstate/clusterstate.go]: done

File: cluster-autoscaler/clusterstate/clusterstate.go (hunk 4 above)
Comment (krzysztof-jastrzebski) [reply cluster-autoscaler/clusterstate/clusterstate.go]: changed

File: cluster-autoscaler/clusterstate/clusterstate.go (hunk 6 above)
Comment (krzysztof-jastrzebski) [reply cluster-autoscaler/clusterstate/clusterstate.go]: done

File: cluster-autoscaler/clusterstate/clusterstate_test.go (hunk 7 above)
Comment (krzysztof-jastrzebski) [reply cluster-autoscaler/clusterstate/clusterstate_test.go]: done

File: cluster-autoscaler/clusterstate/clusterstate.go (hunk 9 above)
Comment (krzysztof-jastrzebski) [reply cluster-autoscaler/clusterstate/clusterstate.go]: done

**Comment by krzysztof-jastrzebski** [comment]
Comments resolved.

File: cluster-autoscaler/clusterstate/clusterstate.go (hunk 2 above)
Comment (losipiuk) [reply cluster-autoscaler/clusterstate/clusterstate.go]: I was thinking usign Threshold instead of interval. But ignore this. Not important.

Comment (losipiuk) [reply cluster-autoscaler/clusterstate/clusterstate.go]: Nothing here. I edited it a few times. And got to same version as original one :)

File: cluster-autoscaler/main.go (hunk 8 above)
Comment (krzysztof-jastrzebski) [reply cluster-autoscaler/main.go]: done

**Comment by losipiuk** [comment]
/lgtm
/approve
Thanks!

**Pull request merged**